packages/aws-cdk/lib/cdk-toolkit.ts

```diff
@@ -534,6 +541,91 @@ export class CdkToolkit {
     return stacks;
   }
 
+  private stackHasNestedStacks(rootStack: cxapi.CloudFormationStackArtifact) {
+    for (const resourceName in rootStack.template.Resources) {
+      if (rootStack.template.Resources[resourceName].Type === 'AWS::CloudFormation::Stack') {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private async findPhysicalNameForNestedStack(logicalId: string, parentStackName: string | undefined, sdk: ISDK): Promise<string | undefined> {
+    if (!parentStackName) {
+      return undefined;
+    }
+    const stackResources = await this.listNestedStackResources(parentStackName, sdk);
+    return stackResources.find(sr => sr.LogicalResourceId === logicalId)?.PhysicalResourceId;
+  }
+
+  private async listNestedStackResources(parentStackName: string, sdk: ISDK): Promise<CloudFormation.StackResourceSummary[]> {
+    const ret = new Array<CloudFormation.StackResourceSummary>();
+    let nextToken: string | undefined;
+    do {
+      const stackResourcesResponse = await sdk.cloudFormation().listStackResources({
+        StackName: parentStackName,
+        NextToken: nextToken,
+      }).promise();
+      ret.push(...(stackResourcesResponse.StackResourceSummaries ?? []));
+      nextToken = stackResourcesResponse.NextToken;
+    } while (nextToken);
+    return ret;
+  }
+
+  private async getNestedStackTemplates(
+    root: cxapi.CloudFormationStackArtifact, nestedTemplateAssetPath: string, nestedStackLogicalId: string,
+    parentStackName: string | undefined, sdk: ISDK,
+  ): Promise<nestedStackTemplates> {
+    const nestedTemplatePath = path.join(root.assembly.directory, nestedTemplateAssetPath);
+    // CFN generates the nested stack name in the form `ParentStackName-NestedStackLogicalID-SomeHashWeCan'tCompute,
+    // so we get the ARN and manually extract the name.
+    const nestedStackArn = await this.findPhysicalNameForNestedStack(nestedStackLogicalId, parentStackName, sdk);
+    const nestedStackName = nestedStackArn?.slice(nestedStackArn.indexOf('/') + 1, nestedStackArn.lastIndexOf('/'));
+
+    return {
+      updatedNestedStackTemplate: JSON.parse(fs.readFileSync(nestedTemplatePath, 'utf-8')),
+      currentNestedStackTemplate: nestedStackName
+        ? await this.props.cloudFormation.readCurrentNestedTemplate(root, nestedStackName) : { Resources: {} },
+      nestedStackName: nestedStackName? nestedStackName: undefined,
+    };
+  }
+
+  private async addNestedStackToParentTemplate(
+    root: cxapi.CloudFormationStackArtifact, parentTemplate: any, currentParentTemplate: any, parentStackName: string | undefined,
+    nestedStackLogicalId: string, sdk: ISDK,
+  ) {
+    const assetPath = parentTemplate.Resources[nestedStackLogicalId].Metadata['aws:asset:path'];
```

Comment:
This expression (`parentTemplate.Resources[nestedStackLogicalId]`) is repeated a lot in this function. We need to assign it to a local variable.